From 8d6620c840697c78a121ccb74e68841d968e2c60 Mon Sep 17 00:00:00 2001
From: Andoni Morales Alastruey <ylatuya@gmail.com>
Date: Mon, 15 Feb 2010 20:33:21 +0100
Subject: [PATCH 3/4] [dshowvideosink] Prepare window and link graph on set_caps()
 	Actually this step is done in the PAUSED_TO_PLAYING state
 	and the window and the graph are not ready when the first
 	preroll needs to be rendered.

---
 sys/dshowvideosink/dshowvideosink.cpp |   27 +++++++++++----------------
 1 files changed, 11 insertions(+), 16 deletions(-)

diff --git a/sys/dshowvideosink/dshowvideosink.cpp b/sys/dshowvideosink/dshowvideosink.cpp
index 1355b2b..90f3ea6 100644
--- a/sys/dshowvideosink/dshowvideosink.cpp
+++ b/sys/dshowvideosink/dshowvideosink.cpp
@@ -752,19 +752,6 @@ gst_dshowvideosink_start_graph (GstDshowVideoSink *sink)
 
   GST_DEBUG_OBJECT (sink, "Connecting and starting DirectShow graph");
 
-  if (!sink->connected) {
-    /* This is fine; this just means we haven't connected yet.
-     * That's normal for the first time this is called. 
-     * So, create a window (or start using an application-supplied
-     * one, then connect the graph */
-    gst_dshowvideosink_prepare_window (sink);
-    if (!gst_dshowvideosink_connect_graph (sink)) {
-      ret = GST_STATE_CHANGE_FAILURE;
-      goto done;
-    }
-    sink->connected = TRUE;
-  }
-
   hres = sink->filter_graph->QueryInterface(
           IID_IMediaControl, (void **) &control);
 
@@ -1333,6 +1320,17 @@ gst_dshowvideosink_set_caps (GstBaseSink * bsink, GstCaps * caps)
   sink->fakesrc->GetOutputPin()->SetMediaType (&sink->mediatype);
   GST_DEBUG_OBJECT (sink, "Configured output pin media type");
 
+  /* We have configured the ouput pin media type.
+  * So, create a window (or start using an application-supplied
+  * one, then connect the graph */
+  gst_dshowvideosink_prepare_window (sink);
+  if (!gst_dshowvideosink_connect_graph (sink)) {
+    GST_ELEMENT_ERROR (sink, CORE, NEGOTIATION,
+          ("Failed to initialize DirectShow graph with the input caps"), (NULL));
+    return FALSE;
+  }
+  sink->connected = TRUE;
+
   return TRUE;
 }
 
@@ -1386,9 +1384,6 @@ gst_dshowvideosink_show_frame (GstVideoSink *vsink, GstBuffer *buffer)
     return GST_FLOW_ERROR;
   }
 
-  if (!sink->connected)
-    return GST_FLOW_OK;
-
   GST_DEBUG_OBJECT (sink, "Pushing buffer through fakesrc->renderer");
   if (!sink->graph_running){
     retst = gst_dshowvideosink_start_graph(sink);
-- 
1.6.4.msysgit.0

